cranelift/codegen/src/opts.rs

```diff
@@ -0,0 +1,309 @@
+//! Optimization driver using ISLE rewrite rules on an egraph.
+
+use crate::egraph::Analysis;
+use crate::egraph::FuncEGraph;
+use crate::egraph::MemoryState;
+pub use crate::egraph::{Node, NodeCtx};
+use crate::ir::condcodes;
+pub use crate::ir::condcodes::{FloatCC, IntCC};
+pub use crate::ir::immediates::{Ieee32, Ieee64, Imm64, Offset32, Uimm32, Uimm64, Uimm8};
+pub use crate::ir::types::*;
+pub use crate::ir::{
+    dynamic_to_fixed, AtomicRmwOp, Block, Constant, DynamicStackSlot, FuncRef, GlobalValue, Heap,
+    Immediate, InstructionImms, JumpTable, MemFlags, Opcode, StackSlot, Table, TrapCode, Type,
+    Value,
+};
+use crate::isle_common_prelude_methods;
+use crate::machinst::isle::*;
+use crate::trace;
+pub use cranelift_egraph::{Id, NewOrExisting, NodeIter};
+use cranelift_entity::{EntityList, EntityRef};
+use smallvec::SmallVec;
+use std::marker::PhantomData;
+
+pub type IdArray = EntityList<Id>;
+#[allow(dead_code)]
+pub type Unit = ();
+pub type Range = (usize, usize);
+
+pub type ConstructorVec<T> = SmallVec<[T; 8]>;
+
+mod generated_code;
+use generated_code::ContextIter;
+
+struct IsleContext<'a, 'b> {
+    egraph: &'a mut FuncEGraph<'b>,
+}
+
+const REWRITE_LIMIT: usize = 5;
+
+pub fn optimize_eclass<'a>(id: Id, egraph: &mut FuncEGraph<'a>) -> Id {
+    trace!("running rules on eclass {}", id.index());
+    egraph.stats.rewrite_rule_invoked += 1;
+
+    if egraph.rewrite_depth > REWRITE_LIMIT {
+        return id;
+    }
+    egraph.rewrite_depth += 1;
+
+    // Find all possible rewrites and union them in, returning the
+    // union.
+    let mut ctx = IsleContext { egraph };
+    let optimized_ids = generated_code::constructor_simplify(&mut ctx, id);
+    let mut union_id = id;
+    if let Some(mut ids) = optimized_ids {
+        while let Some(new_id) = ids.next(&mut ctx) {
+            if ctx.egraph.subsume_ids.contains(&new_id) {
+                trace!(" -> eclass {} subsumes {}", new_id, id);
+                ctx.egraph.stats.node_subsume += 1;
+                // Merge in the unionfind so canonicalization still
+                // works, but take *only* the subsuming ID, and break
+                // now.
+                ctx.egraph.egraph.unionfind.union(union_id, new_id);
+                union_id = new_id;
+                break;
+            }
+            ctx.egraph.stats.node_union += 1;
+            let old_union_id = union_id;
+            union_id = ctx
+                .egraph
+                .egraph
+                .union(&ctx.egraph.node_ctx, union_id, new_id);
+            trace!(
+                " -> union eclass {} with {} to get {}",
+                new_id,
+                old_union_id,
+                union_id
+            );
+        }
+    }
+    ctx.egraph.subsume_ids.clear();
```

Comment:
Actually, thinking about this has made me realize why both this line (and the `subsume` bool also) don't work: reentrancy of the rewriting! From within the call to `simplify` we can create a new node and enter `optimize_eclass` recursively; I think both a clear-at-beginning and clear-at-end potentially prematurely clear `subsume` directives that should be seen by an outer invocation. The most precise memory-reuse approach is probably to clear it when exiting the top level (based on the reentrancy limiter / depth counter) but in order to make reasoning about this easier I think we can probably just let it live for the duration of the egraph (function) rewrite.

Reply:
Ah, that's a good point; I had been imagining a bool flag on the egraph itself rather than the per-ISLE-invocation context. Happy to revisit this (and actually the whole reentrancy mechanism maybe) in followup work...